src/transformers/models/florence2/modular_florence2.py

```diff
@@ -0,0 +1,1782 @@
+        attn = self.softmax(attn)
+
+        x = (attn @ v).transpose(1, 2).reshape(B_, N, C)
+        x = self.proj(x)
+
+        # merge windows
+        x = x.view(-1, self.window_size, self.window_size, C)
+        x = window_reverse(x, B, self.window_size, Hp, Wp)
+
+        if pad_r > 0 or pad_b > 0:
+            x = x[:, :H, :W, :].contiguous()
+
+        x = x.view(B, H * W, C)
+
+        return x, size
+
+
+class SpatialBlock(nn.Module):
+    def __init__(
+        self,
+        dim: int,
+        num_heads: int,
+        window_size: int,
+        mlp_ratio: float = 4.0,
+        qkv_bias: bool = True,
+        drop_path_rate: float = 0.0,
+        act_layer: nn.Module = nn.GELU,
+        norm_layer: nn.Module = nn.LayerNorm,
+        conv_at_attn: bool = True,
+        conv_at_ffn: bool = True,
+    ):
+        super().__init__()
+
+        drop_path = Florence2DropPath(drop_path_rate) if drop_path_rate > 0.0 else nn.Identity()
+
+        self.conv1 = PreNorm(None, DepthWiseConv2d(dim, 3, 1, 1)) if conv_at_attn else None
+        self.window_attn = PreNorm(
+            norm_layer(dim), WindowAttention(dim, num_heads, window_size, qkv_bias=qkv_bias), drop_path
+        )
+        self.conv2 = PreNorm(None, DepthWiseConv2d(dim, 3, 1, 1)) if conv_at_ffn else None
+        self.ffn = PreNorm(
+            norm_layer(dim), Mlp(in_features=dim, hidden_features=int(dim * mlp_ratio), act_layer=act_layer), drop_path
+        )
+
+    def forward(self, x: torch.Tensor, size):
+        if self.conv1:
+            x, size = self.conv1(x, size)
+        x, size = self.window_attn(x, size)
+
+        if self.conv2:
+            x, size = self.conv2(x, size)
+        x, size = self.ffn(x, size)
+        return x, size
+
+
+class DaViTBlock(nn.Module):
+    def __init__(
+        self,
+        embed_dim: int,
+        num_heads: int,
+        window_size: int,
+        drop_path_rate: float,
+        channel_drop_path_rate: float,
+        qkv_bias: bool,
+        mlp_ratio: float,
+        conv_at_attn: bool,
+        conv_at_ffn: bool,
+        num_groups: int,
+    ):
+        super().__init__()
+        self.spatial_block = SpatialBlock(
+            dim=embed_dim,
+            num_heads=num_heads,
+            window_size=window_size,
+            mlp_ratio=mlp_ratio,
+            qkv_bias=qkv_bias,
+            drop_path_rate=drop_path_rate,
+            conv_at_attn=conv_at_attn,
+            conv_at_ffn=conv_at_ffn,
+        )
+        self.channel_block = ChannelBlock(
+            dim=embed_dim,
+            groups=num_groups,
+            qkv_bias=qkv_bias,
+            drop_path_rate=channel_drop_path_rate,
+            mlp_ratio=mlp_ratio,
+            conv_at_attn=conv_at_attn,
+            conv_at_ffn=conv_at_ffn,
+        )
+
+    def forward(self, x: torch.Tensor, size):
+        x, size = self.spatial_block(x, size)
+        x, size = self.channel_block(x, size)
+        return x, size
+
+
+class DaViT(nn.Module):
```

Comment:
Should we implement DaViT separately to follow the Transformers philosophy of "one model per file"?

Reply:
I would normally lean towards yes, but I'm not sure how that interacts with modular transformers - cc @ArthurZucker

Reply:
It's kinda up to you, if you think DaVit will be used alone, maybe, but otherwise we can have FlorenceVit !